File: tenet-0/shared/bus-go/audit.go
package bus

import (
	"context"
	"encoding/json"
	"fmt"
	"strings"
	"time"
)

// Audit returns the audit-log view. Intended for SecOps callers — the
// underlying SELECT requires `tenet0_secops` permissions on audit_log
// (granted in migration 006). Calling Query/Stream from a Bus connected as
// `tenet0_app` yields a permission error wrapped as "audit: query: ...".
//
// The library does not enforce role separation; the caller chooses the
// PostgresURL that determines effective grants.
func (b *Bus) Audit() *Audit {
	return &Audit{bus: b}
}

// Audit wraps read-only queries against audit_log.
type Audit struct {
	bus *Bus
}

// AuditFilter narrows an Audit.Query or Audit.Stream.
// Zero-value fields are ignored.
type AuditFilter struct {
	Actor    string
	Action   string
	FromTime time.Time
	ToTime   time.Time
	Limit    int // default 1000
}

// AuditEntry is one row from audit_log.
type AuditEntry struct {
	ID         int64
	Actor      string
	Action     string
	Detail     json.RawMessage
	RecordedAt time.Time
}

// Query returns matching audit log entries, newest first.
func (a *Audit) Query(ctx context.Context, f AuditFilter) ([]AuditEntry, error) {
	q, args := buildAuditQuery(f)
	rows, err := a.bus.pool.Query(ctx, q, args...)
	if err != nil {
		return nil, fmt.Errorf("audit: query: %w", err)
	}
	defer rows.Close()

	var entries []AuditEntry
	for rows.Next() {
		var e AuditEntry
		if err := rows.Scan(&e.ID, &e.Actor, &e.Action, &e.Detail, &e.RecordedAt); err != nil {
			return nil, fmt.Errorf("audit: scan: %w", err)
		}
		entries = append(entries, e)
	}
	return entries, rows.Err()
}

// Stream polls the audit log and invokes onEntry for each new entry since
// the last poll. Returns a stop function the caller invokes to cancel.
// Each entry fires at most once (tracked by internal last-seen ID).
//
// onEntry runs synchronously inside the polling goroutine; a slow callback
// blocks the next tick. Offload to a goroutine if needed.
func (a *Audit) Stream(ctx context.Context, interval time.Duration, f AuditFilter, onEntry func(AuditEntry)) (stop func(), err error) {
	if interval <= 0 {
		return nil, fmt.Errorf("audit: stream interval must be positive")
	}

	// Start watching from "now" — don't replay history.
	var lastID int64
	err = a.bus.pool.QueryRow(ctx, `SELECT COALESCE(MAX(id), 0) FROM audit_log`).Scan(&lastID)
	if err != nil {
		return nil, fmt.Errorf("audit: stream init: %w", err)
	}

	watchCtx, cancel := context.WithCancel(ctx)
	done := make(chan struct{})

	go func() {
		defer close(done)
		ticker := time.NewTicker(interval)
		defer ticker.Stop()

		for {
			select {
			case <-watchCtx.Done():
				return
			case <-ticker.C:
				newEntries, newLast, err := a.pollSince(watchCtx, lastID, f)
				if err != nil {
					if watchCtx.Err() != nil {
						return
					}
					a.bus.logger.Warn("audit: stream poll failed", "error", err)
					continue
				}
				for _, e := range newEntries {
					onEntry(e)
				}
				if newLast > lastID {
					lastID = newLast
				}
			}
		}
	}()

	return func() { cancel(); <-done }, nil
}

// pollSince fetches audit entries with id > since that match f, oldest first.
func (a *Audit) pollSince(ctx context.Context, sinceID int64, f AuditFilter) ([]AuditEntry, int64, error) {
	clauses, args := buildAuditClauses(f)
	args = append(args, sinceID)
	clauses = append(clauses, fmt.Sprintf("id > $%d", len(args)))

	limit := f.Limit
	if limit <= 0 {
		limit = 1000
	}
	args = append(args, limit)

	q := fmt.Sprintf(`SELECT id, actor_id, action, detail_json, recorded_at
		FROM audit_log WHERE %s ORDER BY id ASC LIMIT $%d`,
		strings.Join(clauses, " AND "), len(args))

	rows, err := a.bus.pool.Query(ctx, q, args...)
	if err != nil {
		return nil, sinceID, err
	}
	defer rows.Close()

	var entries []AuditEntry
	last := sinceID
	for rows.Next() {
		var e AuditEntry
		if err := rows.Scan(&e.ID, &e.Actor, &e.Action, &e.Detail, &e.RecordedAt); err != nil {
			return nil, last, err
		}
		entries = append(entries, e)
		if e.ID > last {
			last = e.ID
		}
	}
	return entries, last, rows.Err()
}

// buildAuditQuery assembles the SELECT for one-shot Query calls.
func buildAuditQuery(f AuditFilter) (string, []any) {
	clauses, args := buildAuditClauses(f)

	limit := f.Limit
	if limit <= 0 {
		limit = 1000
	}
	args = append(args, limit)

	where := ""
	if len(clauses) > 0 {
		where = "WHERE " + strings.Join(clauses, " AND ")
	}

	q := fmt.Sprintf(`SELECT id, actor_id, action, detail_json, recorded_at
		FROM audit_log %s ORDER BY recorded_at DESC LIMIT $%d`, where, len(args))
	return q, args
}

// buildAuditClauses returns the WHERE-clause fragments and bound args for an
// AuditFilter (without leading "WHERE"). Empty clauses for a zero-value filter.
func buildAuditClauses(f AuditFilter) ([]string, []any) {
	var clauses []string
	var args []any

	if f.Actor != "" {
		args = append(args, f.Actor)
		clauses = append(clauses, fmt.Sprintf("actor_id = $%d", len(args)))
	}
	if f.Action != "" {
		args = append(args, f.Action)
		clauses = append(clauses, fmt.Sprintf("action = $%d", len(args)))
	}
	if !f.FromTime.IsZero() {
		args = append(args, f.FromTime)
		clauses = append(clauses, fmt.Sprintf("recorded_at >= $%d", len(args)))
	}
	if !f.ToTime.IsZero() {
		args = append(args, f.ToTime)
		clauses = append(clauses, fmt.Sprintf("recorded_at < $%d", len(args)))
	}
	return clauses, args
}

File: tenet-0/shared/bus-ts/src/audit.ts
import type { Bus } from "./bus.js";

export interface AuditFilter {
  actor?: string;
  action?: string;
  fromTime?: Date;
  toTime?: Date;
  limit?: number; // Default: 1000.
}

export interface AuditEntry {
  id: number;
  actor: string;
  action: string;
  detail: unknown;
  recordedAt: Date;
}

// Audit wraps read-only queries against audit_log. Requires tenet0_secops
// grants on the configured connection; callers using tenet0_app will see a
// permission error.
export class Audit {
  private readonly bus: Bus;

  constructor(bus: Bus) {
    this.bus = bus;
  }

  async query(f: AuditFilter = {}): Promise<AuditEntry[]> {
    const { clauses, args } = buildClauses(f);
    const limit = f.limit && f.limit > 0 ? f.limit : 1000;
    args.push(limit);
    const where = clauses.length > 0 ? `WHERE ${clauses.join(" AND ")}` : "";
    const sql = `SELECT id, actor_id, action, detail_json, recorded_at
                   FROM audit_log ${where} ORDER BY recorded_at DESC LIMIT $${args.length}`;
    const { rows } = await this.bus.pool.query(sql, args);
    return rows.map(rowToEntry);
  }

  // stream polls audit_log for entries with id greater than the last-seen id
  // and invokes onEntry for each. Returns a stop() function. Starts from the
  // current max id (no history replay).
  async stream(
    intervalMs: number,
    f: AuditFilter,
    onEntry: (e: AuditEntry) => void | Promise<void>,
  ): Promise<() => Promise<void>> {
    if (intervalMs <= 0) {
      throw new Error("audit: stream interval must be positive");
    }
    const { rows } = await this.bus.pool.query(
      `SELECT COALESCE(MAX(id), 0)::bigint AS max_id FROM audit_log`,
    );
    let lastId = Number(rows[0]?.max_id ?? 0);
    let stopped = false;
    let pending: Promise<void> | null = null;

    const poll = async (): Promise<void> => {
      if (stopped) return;
      const { clauses, args } = buildClauses(f);
      args.push(lastId);
      clauses.push(`id > $${args.length}`);
      const limit = f.limit && f.limit > 0 ? f.limit : 1000;
      args.push(limit);
      const sql = `SELECT id, actor_id, action, detail_json, recorded_at
                     FROM audit_log WHERE ${clauses.join(" AND ")}
                    ORDER BY id ASC LIMIT $${args.length}`;
      try {
        const { rows } = await this.bus.pool.query(sql, args);
        for (const row of rows) {
          const entry = rowToEntry(row);
          await onEntry(entry);
          if (entry.id > lastId) lastId = entry.id;
        }
      } catch (err) {
        if (!stopped) console.warn("audit: stream poll failed:", err);
      }
    };

    const timer = setInterval(() => {
      if (pending) return;
      pending = poll().finally(() => {
        pending = null;
      });
    }, intervalMs);

    return async () => {
      stopped = true;
      clearInterval(timer);
      if (pending) await pending;
    };
  }
}

function buildClauses(f: AuditFilter): { clauses: string[]; args: unknown[] } {
  const clauses: string[] = [];
  const args: unknown[] = [];
  if (f.actor) {
    args.push(f.actor);
    clauses.push(`actor_id = $${args.length}`);
  }
  if (f.action) {
    args.push(f.action);
    clauses.push(`action = $${args.length}`);
  }
  if (f.fromTime) {
    args.push(f.fromTime);
    clauses.push(`recorded_at >= $${args.length}`);
  }
  if (f.toTime) {
    args.push(f.toTime);
    clauses.push(`recorded_at < $${args.length}`);
  }
  return { clauses, args };
}

function rowToEntry(row: Record<string, unknown>): AuditEntry {
  return {
    id: Number(row.id),
    actor: String(row.actor_id),
    action: String(row.action),
    detail: row.detail_json,
    recordedAt: row.recorded_at as Date,
  };
}
